docs/rfc-1-staking-contract.adoc

```diff
@@ -185,20 +195,56 @@ is higher than the stake amount on T staking contract.
 required to update their information on T staking contract as soon as possible.
-In practice, with the random beacon disabled, and tBTC v1 slashing the stake
-only in case of a proven fraud that had to be committed by all operators of
+In practice, for Keep, with the random beacon disabled, and tBTC v1 slashing the 
+stake only in case of a proven fraud that had to be committed by all operators of
 ECDSA keep, this approach is acceptable.
 
-An integral part of the staking contract should be a bot monitoring stakes and
-notifying about discrepancies. This is especially important given that the bot
-may need to voluntarily inform about discrepancies for operators that have been
-slashed to zero.
+An integral part of the staking contract should be a bot or process inside an
+off-chain client monitoring stakes and notifying about discrepancies. This is
+especially important given that the bot may need to voluntarily inform about
+discrepancies for operators that have been slashed to zero.
+
+Owner or operator can decrease the legacy contract active stake cached amount on
+T staking contract if no application has authorization higher than the liquid T
+stake. It allows to undelegate from the legacy staking contract while still
+being able to operate in T and earn rewards.
 
 ==== Slashing a stake
 
 Authorized applications can slash or seize a stake. Slash operation decreases
 the stake of an operator and burns slashed tokens. Seize decreases the stake,
-burns 95% of the stake, and awards 5% to the notifier of misbehavior. 
+burns 95% of the stake, and awards up to 5% to the notifier of misbehavior.
+
+To keep stakes synchronized between applications when operators are slashed,
+without the risk of running out of gas, the staking contract queues up slashings
+and let users process the transactions.
+
+When an application slashes one or more operators, it adds them to the slashing
+queue on the staking contract. A queue entry contains the operator's address and
+the amount they're due to be slashed.
+
+When there is at least one operator in the slashing queue, any account can
+submit a transaction processing one or more operators' slashings, and collecting
+a reward for doing so. A queued slashing is processed by updating the operator's
+stake to the post-slashing amount, updating authorized amount for each
+affected application, and notifying all affected applications that
+the operator's authorized stake has been reduced due to slashing. The
+application must then do the necessary adjustments, such as removing the
+operator from the sortition pool or reducing its weight, changing the operator's
+eligibility for rewards, and so forth.
+
+Every application callback executed as a result of slash should have a 250k gas
+limit. Slashing are processed in a FIFO basis, and there is just one function
+exposed by the staking contract allowing to slash one or more operators from the
+head of the queue. Callback failure does not revert the transaction.
```

Comment:
I think one possibility would be to move the problematic slashing operation to the end of the queue and take another one from the front of the queue in the same transaction. We'd need to ensure that the operation moving elements in the queue is really cheap so that the cost of it does not exceed gains from executing the slashing.

This has a disadvantage though - in case the application is misbehaving and no slashes from the queue can be executed they keep growing and clogging the queue. One misbehaving application can render slashing of all other applications impossible to execute. Now, if the application can not properly update its internal state, should we let this problem propagate to other applications? This problem could be partially resolved by having the panic button clearing up the queue from slashings of misbehaving application, but before the panic button can disable application slashing and clear the queue, a lot of bad things can happen. This approach also has the disadvantage that the panic button triggered in other circumstances would let some operators benefit from the timing.

Reply:
I think misbehaving applications shouldn't be able to hold up the slashing, and each operator should be cleared from the queue by giving each application one chance to update. If we don't want to just gloss over errors in updating applications we could have another "problem list" where we move the `(operator, application)` pair of each slashing notification call that fails. The problem list would permit notifications on a pairwise basis (so you can call `problemNotify(operator, applicationA)` and it updates application A but leaves application B not updated) for cleaning up, with an arbitrary number of repeat attempts with the pair only being removed from the list once the call is successful. No need for any kind of rewards or other incentives here, just something to facilitate housekeeping.

In the case of a panic buttoned application, the slashing processing should just skip it without calling or adding the operator+application pair to the problem list. Similarly, if a pair on the problem list contains an application that has been panic buttoned, the pair can be simply removed from the list without making a call to the panic-buttoned application.

Reply:
A separate queue for problematic slashings was also proposed by @vzotova today during the weekly call. I can see this idea working but the more I think about it the less I am convinced it is worth the additional complexity. It is in the best interest of the application requesting the slash to be able to process the slashing request from the queue. If the application can not handle it, the application is only hurting itself. I'd vote for giving just one chance to update the state for each application (with a gas cap) and if the application can not process it, the requested update is ignored.